### DISCUSSION TITLE
Parsing PackageURLs

### DISCUSSION THREAD

**Discussion by DamianReeves** [j-mie6/Parsley#225] [Q&A] [answered]
I'm continuing my journey to parse Package URLs via the [purl spec](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst).

I'm wondering how would I go about the first requirement in these instructions. I know this is simple if the entire string is a package url but what if I want to pass the package url in other text (for example if I have a list of package urls) separated by whitespace (since whitespace can't appear in the URL since it would be URL encoded).

If that question is too broad, how would I parse the last "#" sign in a string?

> To parse a purl string in its components:
> 
> Split the purl string once from right on '#'
> The left side is the remainder
> Strip the right side from leading and trailing '/'
> Split this on '/'
> Discard any empty string segment from that split
> Discard any '.' or '..' segment from that split
> Percent-decode each segment
> UTF-8-decode each segment if needed in your programming language
> Join segments back with a '/'
> This is the subpath

**Comment by j-mie6** [comment, marked as answer]
I would parse the whole thing, including the `#` bit, and then use a `map` to process each URL individually to do the splitting and construct the representation of the URL

**Comment by DamianReeves** [comment]
That makes sense, but if I now have a string that is the full URL, how do I parse into 2 parts, the remainder (left side of last #) and subpath (the right side after the #).

**Reply by j-mie6** [reply]
Well, inside a map that's just a normal scala problem, you can split the string on the last hash or something. You're basically escaping parsley to do that more easily

**Reply by DamianReeves** [reply]
Got it. Thanks
